### PR TITLE
[4.0] Namespace JForm and some base classes

### PR DESCRIPTION
**Pull request by laoneo**
### Summary of Changes
This PR is a first step to namespace JForm and fields and rules. It moves JForm/JFormField/JFormHelper/JFormRule to the namespace Joomla\CMS\Form.

New attributes `addfieldprefix`, `addformprefix`, `addruleprefix` are introduced. They are `addfieldpath`, `addformpath`, `addrulepath` and do allow to add a global namespace prefixes. For example in the article.xml file the categoryedit form field is registered like:
```
<?xml version="1.0" encoding="utf-8"?>
<form>
	<fieldset addfieldprefix="Joomla\Component\Categories\Administrator\Field" >
```

The fields in the components com_content and com_categories are moved to namespace as well to illustrate including fields of other components and modal ones.

This pr is dona against the 4.0-dev branch. When i got merged, then I will port it back to J3. I needed some real use cases to see if I'm on the right track.

### Testing Instructions
- Create an article
- Create a "Single Article" menu item with the new article

### Expected result
Article can be created as the category is a required field and uses the categoryedit form field of com_categories.
The article modal field should open a modal window to select an article.

### Actual result
All is working fine.

### Documentation Changes Required
The new attributes do need to be documented.

### CODE REVIEW

File: administrator/components/com_categories/Field/Categoryedit.php [hunk 1]
```diff
@@ -6,19 +6,21 @@
...
-JFormHelper::loadFieldClass('list');
+FormHelper::loadFieldClass('list');
 
 /**
  * Category Edit field..
  *
  * @since  1.6
  */
-class JFormFieldCategoryEdit extends JFormFieldList
+class Categoryedit extends \JFormFieldList
```
Comment (brianteeman) [inline administrator/components/com_categories/Field/Categoryedit.php]: shouldnt this be 
CategoryEdit?

File: administrator/components/com_categories/Field/Categoryedit.php [hunk 2]
```diff
@@ -39,18 +41,18 @@ class JFormFieldCategoryEdit extends JFormFieldList
...
 	 * Method to attach a JForm object to the field.
 	 *
-	 * @param   SimpleXMLElement  $element  The SimpleXMLElement object representing the <field /> tag for the form field object.
-	 * @param   mixed             $value    The form field value to validate.
-	 * @param   string            $group    The field name group control value. This acts as as an array container for the field.
-	 *                                      For example if the field has name="foo" and the group value is set to "bar" then the
-	 *                                      full field name would end up being "bar[foo]".
+	 * @param   \SimpleXMLElement  $element  The SimpleXMLElement object representing the <field /> tag for the form field object.
+	 * @param   mixed              $value    The form field value to validate.
+	 * @param   string             $group    The field name group control value. This acts as as an array container for the field.
```
Comment (brianteeman) [inline administrator/components/com_categories/Field/Categoryedit.php]: as as an ==> as an

I fixed this elsewhere already

File: administrator/components/com_categories/Field/Categoryparent.php [hunk 3]
```diff
@@ -6,17 +6,20 @@
...
+
+FormHelper::loadFieldClass('list');
 
 /**
  * Category Parent field.
  *
  * @since  1.6
  */
-class JFormFieldCategoryParent extends JFormFieldList
+class Categoryparent extends \JFormFieldList
```
Comment (brianteeman) [inline administrator/components/com_categories/Field/Categoryparent.php]: not CategoryParent ?

File: administrator/components/com_categories/Field/Categoryedit.php (hunk 1 above)
Comment (laoneo) [inline administrator/components/com_categories/Field/Categoryedit.php]: It is not possible to magically detect camelcase as we declare the fields lowercase like as can be seen here https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_banners/models/forms/banner.xml#L36

File: administrator/components/com_categories/Field/Categoryparent.php (hunk 3 above)
Comment (laoneo) [inline administrator/components/com_categories/Field/Categoryparent.php]: See categoryedit

File: administrator/components/com_categories/Field/Categoryedit.php (hunk 1 above)
Comment (brianteeman) [inline administrator/components/com_categories/Field/Categoryedit.php]: Thanks for explaining

**Comment by wilsonge** [comment]
After our conversation last night please un-namespace the text form field, fix conflicts and this is good to merge

File: libraries/src/CMS/Form/Form.php [hunk 4]
```diff
@@ -2167,25 +2205,25 @@ public static function getInstance($name, $data = null, $options = array(), $rep
 
 			if (empty($data))
 			{
-				throw new InvalidArgumentException(sprintf('JForm::getInstance(name, *%s*)', gettype($data)));
+				throw new \InvalidArgumentException(sprintf('Form::getInstance(name, *%s*)', gettype($data)));
 			}
 
 			// Instantiate the form.
-			$forms[$name] = new static($name, $options);
+			$forms[$name] = new Form($name, $options);
```
Comment (mbabker) [inline libraries/src/CMS/Form/Form.php]: Keep this as `new static`, allows potential subclasses to use this.

File: tests/unit/suites/libraries/joomla/form/JFormFieldTest.php [hunk 5]
```diff
@@ -40,9 +40,9 @@ protected function setUp()
 	{
 		parent::setUp();
 
-		require_once JPATH_PLATFORM . '/joomla/form/fields/text.php';
-		require_once JPATH_PLATFORM . '/joomla/form/fields/hidden.php';
-		require_once JPATH_PLATFORM . '/joomla/form/fields/checkboxes.php';
+		JLoader::import('joomla.form.fields.text', JPATH_PLATFORM);
```
Comment (mbabker) [inline tests/unit/suites/libraries/joomla/form/JFormFieldTest.php]: Keep these as `require_once`, we aren't trying to test `JLoader` in our unit tests.

**Comment by wilsonge** [comment]
Nice stuff :)

**Pull request merged**